File: me_daily/2025/10/1023/cf105582G/cf105882G.py
from collections import defaultdict
from typing import List
import sys
input = lambda: sys.stdin.readline().strip()
II = lambda: int(input())
LII = lambda: list(map(int, input().split()))
SI = lambda: input()
LSI = lambda: input().split()


n, a, b = LII()
k1 = n // 2
k2 = n - k1

v1, v2 = [0] * k1, [0] * k2

for i in range(k1):
    m, t = LII()
    v1[i] = m * b - a * t
for i in range(k2):
    m, t = LII()
    v2[i] = m * b - a * t

mp = defaultdict(int)

for mask in range(1 << k1):
    val = 0
    for j in range(k1):
        if mask & (1 << j): val += v1[j]
    mp[val] += 1

ans = 0
for mask in range(1 << k2):
    val = 0
    for j in range(k2):
        if mask & (1 << j): val += v2[j]
    ans += mp[-val]
ans -= 1   # 空集被算了一次
print(ans)

File: me_daily/2025/10/1023/cf105582G/cf105882G.rs
use std::collections::HashMap;
use std::io::{self, Read};

macro_rules! read {
    ($it:ident, $t:ty) => {
        $it.next().unwrap().parse::<$t>().unwrap()
    };

    ($it:ident, str) => {
        $it.next().unwrap().to_string()
    };
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input.split_whitespace();

    let n: usize = read!(it, usize);
    let a: usize = read!(it, usize); // usize，但calc转i64
    let b: usize = read!(it, usize);
    let k1 = n / 2;
    let k2 = n - k1;
    let mut v1: Vec<i64> = vec![0; k1];
    let mut v2: Vec<i64> = vec![0; k2];

    // 第一半：读k1对，赋值v1[i]
    for i in 0..k1 {
        let m: i64 = read!(it, i64);
        let t: i64 = read!(it, i64);
        v1[i] = m * (b as i64) - (a as i64) * t; // cast usize to i64
    }
    // 第二半：读k2对，赋值v2[i]
    for i in 0..k2 {
        let m: i64 = read!(it, i64);
        let t: i64 = read!(it, i64);
        v2[i] = m * (b as i64) - (a as i64) * t;
    }

    let mut mp: HashMap<i64, usize> = HashMap::new();
    // 第一半mask：枚举所有子集sum
    for mask in 0..(1usize << k1) {
        let mut val: i64 = 0i64;
        for j in 0..k1 {
            if (mask & (1usize << j)) != 0 {
                val += v1[j];
            }
        }
        *mp.entry(val).or_insert(0) += 1;
    }

    let mut ans: usize = 0; // 加定义！
                            // 第二半mask：枚举，查mp[-val]
    for mask in 0..(1usize << k2) {
        let mut val: i64 = 0i64;
        for j in 0..k2 {
            if (mask & (1usize << j)) != 0 {
                val += v2[j];
            }
        }
        if let Some(&cnt) = mp.get(&-val) {
            ans += cnt;
        }
    }
    println!("{}", ans.saturating_sub(1)); // -1，但防usize负（虽ans>=1）
}
